### ISSUE TITLE
`TypeVar` inference retains `Literal` types instead of widening to primitive types in generic calls

### ISSUE THREAD

**Issue by cmp0xff** [astral-sh/ty#4363]
### Summary

https://play.ty.dev/fea6b83d-e90a-40f1-b9a2-3753b0db9770

```py
from typing import TypeVar, assert_type
    
T = TypeVar("T")
    
def get(default: T) -> int | T:
    return default
    
# In ty: inferred as int | Literal["a"]
# In mypy/pyrefly: inferred as int | str
assert_type(get("a"), int | str)
```

## PEP 586 Specification Violation

1. Default Widening: PEP 586 § Type Inference https://peps.python.org/pep-0586/#type-inference requires type checkers to infer primitive base types (str, int, bool) for literal expressions by default, unless the parameter or TypeVar is explicitly constrained to Literal.
2. TypeVar Solving: When binding an unannotated literal argument to a generic function parameter T, T should solve to the widened type str rather than over-specializing to Literal["a"].
3. Ecosystem Consistency: Standard type checkers (mypy, pyrefly, pyright) widen literal arguments in this context to ensure expected behavior with generic signatures (such as dict.get or Series.get) and prevent invariance issues in generic containers.

Co-authored-by: Antigravity AI

### Version

playground

**Comment by carljm** [comment]
PEP 586 specifically says that "no inference strategy is mandated" and encourages type checkers to experiment with more sophisticated inference strategies.

None of the downsides of inferring literal types discussed in PEP 586, or in this issue, apply to ty's inference strategy. ty does always promote literal types in the context of an invariant generic; `[get("a")]` is inferred in ty as `list[int | str]`. And (unlike mypy, which formed the context for the recommendation in PEP 586), ty never infers an upper bound for a variable's type without an explicit annotation, so inferring `x = get("a")` as `int | "a"` in ty never prevents later assigning `"b"` to `x`.

So for ty there is no downside to inferring `get("a")` as `int | Literal["a"]` -- it is a more precise type that is always accurate, and does not impose any additional constraints on the user's code. (On the contrary, it can avoid false positives if that value is later passed somewhere requiring a specific literal type.)
